### PR TITLE
Forkchoice external locks

### PR DESCRIPTION
**Pull request by potuz**
# Forkchoice reentrancy

## TL; DR:

Replace all locks internal to forkchoice by a single `forkchoice.Lock()`  that is acquired by the highest level caller possible. This avoids all race conditions and is much less bugprone

## Rationale

Forkchoice has many calls that modify its internal structure:

- Updating the current head, on calls to `Head()` by recomputing the nodes balances.
- Inserting a new node or chain of nodes, on calls to `InsertNode()` and `InsertChain()`.
- Inserting new attestations or slashed indices on calls to `ProcessAttestation()` and `InsertSlashedIndex()`.
- Updating the current slot, called at the beginning of each slot on calls to `NewSlot()`. This call updates the checkpoints and prunes old nodes.
- Updating the optimistic status of a node 

Other setters only happen at node start like setting the checkpoints, the balance by root handler, the origin root (when starting from genesis), etc. 

In addition, Forkchoice has many getters that read its internal structure. These getters request if a root is known to fork choice, the canonical status of a node, the current checkpoints, etc. 

## Setters

The setters, those that will require a write lock, are called in the following scenarios:

- On block processing: `OnBlock` updates the head, the optimistic status of the newly inserted node, adds attestations from the block, inserts slashed indices. In addition, this path requires several reads from forkchoice: justified checkpoints, weights (for reporting in case of reorgs), etc. 
- On attestation processing: This path is triggered at the start of each slot, with a call to `NewSlot()` on a time ticker every 12 seconds, and then calls to `UpdateHead()`. This path is also called when proposing a block, right before proposing. 

## Getters

Getters can be triggered at any time from RPC calls, most calls would require at least to read the head and its optimistic status, and some will require the finalized and justified checkpoints. 

## Lock contention

Even disregarding getters, that access very specific structures in forkchoice, we need to have write logs because block processing and attestation processing can happen at the same time (a block may arrive at 12 seconds exactly). However,
the current interlocking of forkchoice is useless: block processing acquires and releases several times locks from forkchoice and if processing a block at the same time of `NewSlot()` the justified checkpoints may become inconsistent during the block processing. 

With the current interlocking, a single block that arrives and is processed will require multiple locks to be acquired and released consecutively, even within each getter and setter called there are several locks being acquired/released. The current justified and finalized checkpoint may change between one call and the next one, rendering the whole process fragile. On the other hand, there is nothing to gain in allowing an RPC call to succeed while we are processing a block, there is one situation where this is not the case: when proposing a block, we would want to do it even if we can't acquire the lock for forkchoice.

We propose a single lock to forkchoice, controlled by the external caller to forkchoice.  Setters will acquire a write lock and getters will acquire a read lock. Given that setters only get called during either block processing or updating head, and in any case it's preferable to run the whole procedure with a locked forkchoice, it is simple to acquire a write lock when starting to process a block and release it when done. Similarly on new slots and block proposals. 

For getters the situation is slightly more complicated. We do not want to have long running RPC calls locking forkchoice for the whole duration of the RPC call, so care must be taken to acquire and release the forkchoice lock immediately.

With this proposal, all of forkchoice becomes single threaded and there is no need to have internal locks, the callers, which are higher level functions, are required to acquire a single lock to interact with Forkchoice. This is less bug prone than the current approach as the number of calls to forkchoice is minimal and there aren't many nested calls. 

`IsOptimistic()` from `chain_info` is called in `saveHead`,  `notifiedNewHeadEvent` and `Status()`, and `saveHead()` is called when calling FCU, which in turn is in the block processing path. 
## Concrete implementation

- Remove all locks from within the `doublylinkedtree` package.
- Global write lock on `ReceiveBlock` and `ReceiveBlockBatch`.
- write locks on `GetBeaconBlock` and on the ticker for `NewSlot()`.
- Remove `processAttestationLock` as it's no longer needed.
- Read Lock on `blockchain.Status()` (called from the `healthz` endpoint).
- Read Lock on all (g)RPC methods that call forkchoice downstream. 

## Notes

- The function `CachedHeadRoot()` is currently unused, but we think it may be worth keeping it for future optimizations in which we don't recompute head in certain situations and remove the responsibility of blockchain in keeping head information. 

- There is one instance of nested locks in that `ProduceBlockV2` calls `ValidateSync` and then calls `GetBeaconBlock` that requires a write lock due to updating head. I feel that there is no need to validate the syncing status since `GetBeaconBlock` will do that anyway.

- [x] Need to audit that there are no recursive write locks
- [ ] Need to audit that there are no recursive locks at all. 
- [x] Need to finish with RPC methods in `rpc/statefetcher` That are called from `rpc/eth/validator` and they themselves require the finalized checkpoint. 
- [x] Need to audit the stategen package that has now access to forkchoice and it may be that `StateByRoot()` or similar require forkchoice somehow.

### CODE REVIEW

File: beacon-chain/blockchain/head.go [hunk 1]
```diff
@@ -158,12 +158,16 @@ func (s *Service) saveHead(ctx context.Context, newHeadRoot [32]byte, headBlock
 
 	// Forward an event capturing a new chain head over a common event feed
 	// done in a goroutine to avoid blocking the critical runtime main routine.
-	go func() {
-		if err := s.notifyNewHeadEvent(ctx, newHeadSlot, headState, newStateRoot[:], newHeadRoot[:]); err != nil {
-			log.WithError(err).Error("Could not notify event feed of new chain head")
-		}
-	}()
-
+	optimistic, err := s.IsOptimistic(ctx)
```
Comment (potuz) [inline beacon-chain/blockchain/head.go:161]: moving the check here so that the lock to forkchoice required by `IsOptimistic` is not acquired on a go routine.

File: beacon-chain/blockchain/receive_block.go [hunk 2]
```diff
@@ -45,6 +45,9 @@ func (s *Service) ReceiveBlock(ctx context.Context, block interfaces.ReadOnlySig
 		return err
 	}
 
+	s.cfg.ForkChoiceStore.Lock()
```
Comment (potuz) [inline beacon-chain/blockchain/receive_block.go:48]: Single global lock for all of block processing

File: beacon-chain/blockchain/receive_block.go [hunk 3]
```diff
@@ -90,6 +93,9 @@ func (s *Service) ReceiveBlockBatch(ctx context.Context, blocks []interfaces.Rea
 	ctx, span := trace.StartSpan(ctx, "blockChain.ReceiveBlockBatch")
 	defer span.End()
 
+	s.cfg.ForkChoiceStore.Lock()
```
Comment (potuz) [inline beacon-chain/blockchain/receive_block.go:96]: Single global lock for all of batch processing. Need to check that `ReceiveBlock` is not called from within `ReceiveBlockBatch`

File: beacon-chain/forkchoice/doubly-linked-tree/types.go [hunk 4]
```diff
@@ -11,12 +11,12 @@ import (
 
 // ForkChoice defines the overall fork choice store which includes all block nodes, validator's latest votes and balances.
 type ForkChoice struct {
+	lock                sync.RWMutex // global forkchoice lock
```
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/types.go:14]: If you do this then you do not need to redefine the lock and unlock methods

```suggestion
	sync.RWMutex // global forkchoice lock
```

File: beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go [hunk 5]
```diff
@@ -720,3 +647,23 @@ func (f *ForkChoice) updateJustifiedBalances(ctx context.Context, root [32]byte)
...
+
+// RLock acquires the global read lock
+func (f *ForkChoice) RLock() {
+	f.lock.RLock()
+}
+
+// RUnlock releases the global read lock
+func (f *ForkChoice) RUnlock() {
+	f.lock.RUnlock()
+}
```
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go:669]: You don't need to define these by using strictly `sync.RWMutex` without a field name in `sync.RWMutex`

```suggestion
```

File: beacon-chain/forkchoice/interfaces.go [hunk 6]
```diff
@@ -16,13 +16,22 @@ type BalancesByRooter func(context.Context, [32]byte) ([]uint64, error)
...
+	ForkChoiceLocker     // to lock forkchoice
 	HeadRetriever        // to compute head.
 	BlockProcessor       // to track new block for fork choice.
 	AttestationProcessor // to track new attestation for fork choice.
 	Getter               // to retrieve fork choice information.
 	Setter               // to set fork choice information.
 }
 
+// ForkChoiceLocker defines an interface for methods that aquire and release the forkchoice locks
+type ForkChoiceLocker interface {
```
Comment (terencechain) [inline beacon-chain/forkchoice/interfaces.go:28]: `forkchoice.ForkChoiceLocker` is verbose, how about just `forkchoice.Locker` ?

```suggestion
type Locker interface {
```

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 7]
```diff
@@ -41,6 +41,8 @@ func (vs *Server) GetBeaconBlock(ctx context.Context, req *ethpb.BlockRequest) (
 	ctx, span := trace.StartSpan(ctx, "ProposerServer.GetBeaconBlock")
 	defer span.End()
 	span.AddAttributes(trace.Int64Attribute("slot", int64(req.Slot)))
+	vs.ForkChoiceLocker.Lock()
+	defer vs.ForkChoiceLocker.Unlock()
```
Comment (terencechain) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:45]: I feel a bit strongly about reducing the surface area for this lock. We should be able to release it after `HeadState`. Is there a reason why we are not releasing it earlier

Comment (terencechain) [reply beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:45]: To answer my question, simplicity and clarity is one reasons not to do it

**Review by terencechain — CHANGES_REQUESTED** [review]
~~We are missing many locks in the `sync` package as part of the validation pipeline. Please double-check the `HeadState` usages there~~

EDIT: looking further, I think this is safe. head state from blockchain service package is not from forkchoice

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go (hunk 7 above)
Comment (terencechain) [reply beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:45]: Also `circuitBreakBuilder` uses head fetcher as well. I think it's best to keep it as it is

File: beacon-chain/rpc/eth/beacon/blinded_blocks.go [hunk 8]
```diff
@@ -24,6 +24,8 @@ import (
 func (bs *Server) GetBlindedBlock(ctx context.Context, req *ethpbv1.BlockRequest) (*ethpbv2.BlindedBlockResponse, error) {
 	ctx, span := trace.StartSpan(ctx, "beacon.GetBlindedBlock")
 	defer span.End()
+	bs.ChainInfoFetcher.ForkChoicer().RLock()
+	defer bs.ChainInfoFetcher.ForkChoicer().RUnlock()
```
[the remark this answers is not in the document]
Comment (potuz) [reply beacon-chain/rpc/eth/beacon/blinded_blocks.go:28]: Yeah there's a danger here but in principle only the highest level API calls need a lock

**Comment by terencechain** [comment]
> ~We are missing many locks in the `sync` package as part of the validation pipeline. Please double-check the `HeadState` usages there~
> 
> EDIT: looking further, I think this is safe. head state from blockchain service package is not from forkchoice

Something unsafe is the usage of `s.cfg.chain.FinalizedCheckpt()` in rpc status, blocks fetcher, pending queue and round robins. Those will need to be locked

File: beacon-chain/rpc/eth/beacon/blinded_blocks.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/rpc/eth/beacon/blinded_blocks.go:28]: Maybe the highest level here is just any beacon API call. Not on the function level. This is the safest but also not efficient

**Comment by potuz** [comment]
Closing in favor of #12036
